demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,613 @@
+            {
+                configPRINTF( ( "Error getting attribute value of EC key type.\r\n" ) );
+            }
+
+            if( xTemplate.ulValueLen != sizeof( CK_KEY_TYPE ) )
+            {
+                configPRINTF( ( "Length of key type incorrect in GetAttributeValue.\r\n" ) );
+            }
+
+            if( xKeyType != CKK_EC )
+            {
+                configPRINTF( ( "Incorrect key type for public key.\r\n" ) );
+            }
+
+            /* Check that public key point can be retrieved for public key. */
+            xTemplate.type = CKA_EC_POINT;
+            xTemplate.pValue = xEcPoint;
+            xTemplate.ulValueLen = sizeof( xEcPoint );
+            xResult = pxGlobalFunctionList->C_GetAttributeValue( xGlobalSession, xPublicKeyHandle, &xTemplate, 1 );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to retrieve EC Point.\r\n" ) );
+            }
+
+            /* Perform a sign with the generated private key. */
+            xMechanism.mechanism = CKM_ECDSA;
+            xMechanism.pParameter = NULL;
+            xMechanism.ulParameterLen = 0;
+            xResult = pxGlobalFunctionList->C_SignInit( xGlobalSession, &xMechanism, xPrivateKeyHandle );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to SignInit ECDSA.\r\n" ) );
+            }
+
+            xSignatureLength = sizeof( xSignature );
+            xResult = pxGlobalFunctionList->C_Sign( xGlobalSession, xHashedMessage, SHA256_DIGEST_SIZE, xSignature, &xSignatureLength );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to ECDSA Sign.\r\n" ) );
+            }
+
+            /* Verify the signature with mbedTLS */
+            mbedtls_ecdsa_init( &xEcdsaContext );
+            mbedtls_ecp_group_init( &xEcdsaContext.grp );
+
+
+            lMbedTLSResult = mbedtls_ecp_group_load( &xEcdsaContext.grp, MBEDTLS_ECP_DP_SECP256R1 );
+
+            if( lMbedTLSResult != 0 )
+            {
+                configPRINTF( ( "mbedTLS failed in setup for signature verification.\r\n" ) );
+            }
+
+            /* The first 2 bytes are for ASN1 type/length encoding. */
+            lMbedTLSResult = mbedtls_ecp_point_read_binary( &xEcdsaContext.grp, &xEcdsaContext.Q, &xEcPoint[ 2 ], xTemplate.ulValueLen - 2 );
+
+            if( lMbedTLSResult != 0 )
+            {
+                configPRINTF( ( "mbedTLS failed in setup for signature verification.\r\n" ) );
+            }
+
+            /* Generate a PK Context for CSR */
+            /* Getting header info from ECKEY type */
+            mbedtls_pk_type_t type_key = MBEDTLS_PK_ECKEY;
+            const mbedtls_pk_info_t * header = mbedtls_pk_info_from_type( type_key );
+
+            /* Creating copy of header to pass in custom sign function (original header is immutable) */
+            mbedtls_pk_info_t * header_copy = pvPortMalloc( sizeof( mbedtls_pk_info_t ) );
```

Comment:
Need to check for resource exhaustion.

Reply:
Added Null checks for malloc calls. Thanks for the pointer!